src/myrtlespeech/model/rnn.py

```diff
@@ -103,49 +112,93 @@ def __init__(
-                sequence.
+                in_features]``. ``hid`` is the RNN hidden state which is
+                either a length 2 Tuple of :py:class:`torch.Tensor`s or
+                a single :py:class:`torch.Tensor` depending on the ``RNNType``
+                (see :py:class:`torch.nn` documentation for more information).
+
+                The return type `res[0]` will be the same as the `x[0]` type so
+                you should pass `hid = None` if you would like the hidden state
+                returned and it is the start-of-sequence. In this case, the
+                hidden state(s) will be initialised to zero in PyTorch.
+
+                ``x[1]`` is a :py:class:`torch.Tensor` where each entry
+                represents the sequence length of the corresponding network
+                *input* sequence.
 
         Returns:
-            The first element of the Tuple return value is the result after
-            applying the RNN to ``x[0]``. It must have size ``[seq_len,
-            batch, out_features]`` or ``[batch, seq_len, out_features]`` if
-            ``batch_first=True``.
-
-            The second element of the Tuple return value is a
-            :py:class:`torch.Tensor` with size ``[batch]`` where each entry
-            represents the sequence length of the corresponding *output*
-            sequence. This will be equal to ``x[1]`` as this layer does not
-            currently change sequence length.
+            A Tuple ``(res[0], res[1])``. ``res[0]`` will take the same form as
+            ``x[0]``: either a tuple ``res[0] = (out, hid)`` or a
+            :py:class:`torch.Tensor``. ``res[0] = out``. ``out`` is the
+            result after applying the RNN to ``inp``. It will have size
+            ``[seq_len, batch, out_features]``. ``hid`` is the
+            returned RNN hidden state which is either a length 2 Tuple of
+            :py:class:`torch.Tensor`s or a single :py:class:`torch.Tensor`
+            depending on the ``RNNType`` (see :py:class:`torch.nn`
+            documentation for more information).
+
+            ``res[1]`` is a :py:class:`torch.Tensor` where each entry
+            represents the sequence length of the corresponding network
+            *output* sequence. This will be equal to ``x[1]`` as this layer
+            does not change sequence length.
         """
+
+        if isinstance(x[0], torch.Tensor):
+            inp = x[0]
+            hid = None
+            return_tuple = False
+        elif isinstance(x[0], tuple) and len(x[0]) == 2:
+            inp, hid = x[0]
+            return_tuple = True
+        else:
+            raise ValueError(
+                "`x[0]` must be of form (input, hidden) or (input)."
+            )
+
         if self.use_cuda:
-            x = (x[0].cuda(), x[1].cuda())
+            inp = inp.cuda()
+            if hid is not None:
+                if isinstance(hid, tuple):  # LSTM/GRU
+                    hid = hid[0].cuda(), hid[1].cuda()
```

Comment:
Yes v. good point - I've added a check to ensure the hidden state is of length 2!

To your first point: I don't think we want to allow more general hidden states because, if `pytorch` adds this in the future we should force anyone updating `myrtlespeech` to think explicitly about hidden state parsing. Concretely, after the BN PR goes in, each rnn layer will be a new `torch.nn.LSTM`/GRU and hence we will be using helper functions to gather and arrange the hidden state for the `hx=(h_0, c_0)` and `hx=h_0`.